be/src/common/status.h

```diff
@@ -17,114 +17,136 @@ namespace doris {
 
 class Status {
 public:
-    Status(): _state(nullptr) {}
+    Status() : _state(nullptr) { }
     ~Status() noexcept { delete[] _state; }
 
-    // copy c'tor makes copy of error detail so Status can be returned by value
-    Status(const Status& s)
-            : _state(s._state == nullptr ? nullptr : copy_state(s._state))  {
-    }
+    // Here _state == nullptr(Status::OK) is the most common case, so use this as
+    // a fast path. This check can avoid one function call in most cases, because
+    // _copy_state() is not an inline function.
+    Status(const Status& s) : _state(s._state == nullptr ? nullptr : _copy_state(s._state))  { }
 
     // same as copy c'tor
     Status& operator=(const Status& s) {
         // The following condition catches both aliasing (when this == &s),
         // and the common case where both s and *this are OK.
         if (_state != s._state) {
             delete[] _state;
-            _state = (s._state == nullptr) ? nullptr : copy_state(s._state);
+            _state = (s._state == nullptr) ? nullptr : _copy_state(s._state);
         }
         return *this;
     }
 
-    // move c'tor
     Status(Status&& s) noexcept : _state(s._state) {
         s._state = nullptr;
     }
 
-    // move assign
     Status& operator=(Status&& s) noexcept {
-        std::swap(_state, s._state);
+        if (_state != s._state) {
```

Comment:
what's wrong with std::swap?

Reply:
Taken from [https://stackoverflow.com/questions/6687388/why-do-some-people-use-swap-for-move-assignments](https://stackoverflow.com/questions/6687388/why-do-some-people-use-swap-for-move-assignments)

Reply:
For your example, is it OK to use an object after it has been moved?

Reply:
Yes, it is ok to use a variable after `std::move()` it. 
At least the compiler will not prevent anyone from using it this way.

Reply:
What I am worried about is that the C ++ standard requires that it must be initialized after the move?

Reply:
No, I do not think the standard required that. But for `dest = std::move(src)`, the following two points are determined:
1. the `dest` resource should **not** be assigned to `src` after move-assignment.
2. The move-assignment of move-constructor should have the **same behavior**.

And for `Status`, it only has one member, `_state`, which is **pointer type**.

when the `src::_state` was moved to `dest`, it must be cleared in `src`, otherwise, the pointer will be referenced by both the `src` and `dest`, which will be wrong.